### PR TITLE
KAFKA-4427: Skip topic groups with no tasks

### PR DESCRIPTION
**Pull request by enothereska**

### CODE REVIEW

**Comment by enothereska** [comment]
@dguy @guozhangwang please have a look if you can. Thanks.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 1]
```diff
@@ -430,15 +430,18 @@ public Subscription subscription(Set<String> topics) {
...
-                    if (numPartitions < task.partition + 1)
-                        numPartitions = task.partition + 1;
+                if (tasksByTopicGroup.get(topicGroupId) != null) {
+                    for (TaskId task : tasksByTopicGroup.get(topicGroupId)) {
+                        if (numPartitions < task.partition + 1)
+                            numPartitions = task.partition + 1;
+                    }
                 }
+                if (numPartitions > -1) {
+                    InternalTopicMetadata topicMetadata = new InternalTopicMetadata(topicConfig);
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:438]: Any reason why this block of code isn't in the `if(tasksByTopicGroup.get(topicGroupId) != null)` block? After the `for(..)`?
 If it is not null then it is going to have some tasks, right? So numPartitions will always be > -1

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignorTest.java [hunk 2]
```diff
@@ -215,6 +215,55 @@ public void testAssignBasic() throws Exception {
...
+        TopologyBuilder builder = new TopologyBuilder();
+        builder.addSource("source1", "topic1");
+        builder.addProcessor("processor1", new MockProcessorSupplier(), "source1");
+        builder.addStateStore(new MockStateStoreSupplier("store1", false), "processor1");
+        builder.addSource("source2", "topic2");
+        builder.addProcessor("processor2", new MockProcessorSupplier(), "source2");
+        builder.addStateStore(new MockStateStoreSupplier("store2", false), "processor2");
+        List<String> topics = Utils.mkList("topic1", "topic2");
+
+        final Set<TaskId> prevTasks10 = Utils.mkSet(task0);
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignorTest.java]: I don't think we need these prevTasks and standbyTasks for this test. You can just pass `Collections.emptySet()` to the `Subscription`

**Review by dguy — COMMENTED** [review]
Couple of minor comments

**Comment by enothereska** [comment]
Thanks @dguy

**Review by dguy — APPROVED** [review]
LGTM

**Comment by enothereska** [comment]
org.apache.kafka.clients.consumer.KafkaConsumerTest.testCloseTimeout failed. This is unrelated to this PR.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/SingleGroupPartitionGrouperStub.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.processor.internals;
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/SingleGroupPartitionGrouperStub.java:18]: Should we move this class file under `test`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 4]
```diff
@@ -430,15 +430,16 @@ public Subscription subscription(Set<String> topics) {
...
-                        numPartitions = task.partition + 1;
-                }
-
-                InternalTopicMetadata topicMetadata = new InternalTopicMetadata(topicConfig);
-                topicMetadata.numPartitions = numPartitions;
+                if (tasksByTopicGroup.get(topicGroupId) != null) {
+                    for (TaskId task : tasksByTopicGroup.get(topicGroupId)) {
+                        if (numPartitions < task.partition + 1)
+                            numPartitions = task.partition + 1;
+                    }
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:437]: Could we log a debug entry indicating some sub-topology's tasks was not created?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignorTest.java [hunk 5]
```diff
@@ -215,6 +215,38 @@ public void testAssignBasic() throws Exception {
...
+
+        StreamPartitionAssignor partitionAssignor = new StreamPartitionAssignor();
+        partitionAssignor.configure(config.getConsumerConfigs(thread10, "test", client1));
+        Map<String, PartitionAssignor.Subscription> subscriptions = new HashMap<>();
+        subscriptions.put("consumer10",
+            new PartitionAssignor.Subscription(topics, new SubscriptionInfo(uuid1, Collections.<TaskId>emptySet(), Collections.<TaskId>emptySet(), userEndPoint).encode()));
+
+        // will throw exception if it fails
+        partitionAssignor.assign(metadata, subscriptions);
+    }
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignorTest.java:254]: Checks the number of tasks as well?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignorTest.java [hunk 6]
```diff
@@ -215,6 +215,38 @@ public void testAssignBasic() throws Exception {
     }
 
     @Test
+    public void testAssignBasicSingleGroupPartitionGrouper() throws Exception {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignorTest.java]: nit: `testAssignWithPartialTopology`

**Review by guozhangwang — COMMENTED** [review]
LGTM overall, minor comments.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/SingleGroupPartitionGrouperStub.java (hunk 3 above)
Comment (enothereska) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/SingleGroupPartitionGrouperStub.java:18]: @guozhangwang it is already in `src/test` with the same package that other similar files belong to (e.g., see `SinkNodeTest`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java (hunk 4 above)
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:437]: Done

**Comment by enothereska** [comment]
kafka.producer.ProducerTest.testAsyncSendCanCorrectlyFailWithTimeout failed. Unrelated to PR. Streams tests passed.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/SingleGroupPartitionGrouperStub.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/SingleGroupPartitionGrouperStub.java:18]: That's right. My bad.

**Comment by guozhangwang** [comment]
Merged to trunk.

**Pull request closed without merging**